### ISSUE TITLE
assign weights/number of workers for each hosts

### ISSUE THREAD

**Issue by minhhg** [uqfoundation/pathos#71]
Hi,
It is not clear how to assign weights or number of workers for each hosts.  I got a slow laptop and 2 servers but the server pool seems to give the same weight for all. Is there any way I can assign,divide jobs by myself.

**Comment by mmckerns** [comment]
@minhhg: Let me understand the situation a bit better… you have a slow laptop, with some local workers, and two remote servers that you are connecting with.  If you are using `pathos.pools.ProcessPool`, any `map` you choose will pull from the pool of workers by gauging your internet connection and the local processors available.  If you want something other than that, say to ensure you always have 9 of 10 workers on remote servers, then that's not the worker pool strategy, and you'd have to code something to that effect. 

For example, you could split your jobs into 3 bins, and then do a pool for each of the different bins.  That would run a pool on each resource.  It's possible (I've done it before), to build a hierarchy of `map` functions, one level to split the jobs up and to start pools for the individual resources, and the second level to run the pools for each resource.  You could then, theoretically, build something to distribute your jobs as you see fit.  I don't provide this at a high-level configurable strategy, like `pool`.  The second-most common strategy is the "equipartition" strategy, which splits up all the jobs into equal parts and submits each portion to a worker.  This is fairly straight-forward to implement with a nested pool strategy, but that still doesn't seem to be exactly what you are looking for if you are asking about weights.

In short, `pathos` does not provide custom strategies for distributing jobs -- just the worker pool strategy.  However, you can build on that to create your own custom strategy.

If this is a feature request for a certain high-level strategy, please describe it more fully.

**Comment by minhhg** [comment]
Hi mmckerns,
Thanks for the suggestions. I just thought that would be usefull to have something like that. However I think that may not be neccessary until the ppmap and pp_map or  pp.submit work flasslessly.
Currently I have a code, which works fine with pathos.multiprocessing map but if I switches  to use the other ones, I got the following error message (see below). I want to run on l "cluster" using tunnel.
Is that possible to make it works? I thought this is the main aim of pathos. 
----------------------------------------------------------------
(in my code I use nplopt package)
An error has occured during the function import
Traceback (most recent call last):
  File "/home/xxx/miniconda/envs/banks/lib/python2.7/site-packages/ppft-1.6.4.5-py2.7.egg/ppworker.py", line 82, in run
    exec __fobj
  File "<string>", line 79, in <module>
  File "<string>", line 85, in SwigPyIterator
NameError: name '_swig_repr' is not defined
A fatal error has occured during the function execution
Traceback (most recent call last):
  File "/home/xxx/miniconda/envs/banks/lib/python2.7/site-packages/ppft-1.6.4.5-py2.7.egg/ppworker.py", line 91, in run
    __f = locals()[__fname]
KeyError: 'do_optimize'

**Comment by mmckerns** [comment]
@minhhg: It's definitely useful to have custom strategies, or at least different strategies for distributing work.  It's a high-level programming task around the low-level functionality that's currently in the `pathos`, and it's not there primarily because I haven't had a serious need for it yet.  It would be the most useful in conjunction with `ppft` in distributing jobs on different clusters, but (again) I have only run that way infrequently -- more commonly, I use `pathos` in conjunction with (my MPI map package) `pyina` to distribute jobs in a hierarchical fashion over large institutional class compute clusters using a scheduler or schedulers.  `pyina` does have different job distribution strategies.  So, I absolutely agree that it's not a bad suggestion/request… it's just not implemented in `pathos` just yet.

With regard to your error above, it looks like your function is failing to pickle.  The `ppft` package pickles by extracting the source code of objects and then rebuilding them on the target machine.  First off, using SWIG typically means C/C++, which can't pickle… secondly, iterators only pickle in python `3.4`+ thus far (but would work if the source was pure python). The `KeyError` you are seeing means that `ppft` is trying to look up the `do_optimize` function in the `locals` dictionary , which holds the source codes of the functions being "pickled", and it's not finding it.  You can pass reference modules to help the serialization (or in this case, code extraction), but I don't think that's your problem.  You might want to put your target function in a module that can be found on both your local and target computer… and then it will be possible to get around limitations of having SWIG and iterations in your function call.

**Comment by mmckerns** [comment]
please feel free to post here if this warrants reopening
